lib/trace/core.js

```diff
@@ -0,0 +1,108 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var events = require('events');
+var util = require('util');
+
+var _ = require('underscore');
+
+var tracerConfigMap = require('./config');
+
+// resolves and returns a tracer config object, with
+// extEvent, sourceEmitter, sourceEvent
+function resolveTracerConfig(ringpop, extEvent) {
+    var tracerConfig = tracerConfigMap[extEvent];
+    var emitterPath = tracerConfig.sourcePath;
+
+    // XXX subtle but important -- if we resolve here, then the
+    // underlying object can never change. if we do latent resolution,
+    // there may be a resource leak when we try to remove listener
+    // from something that is out of our reachable scope.
+    var emitter = getIn(ringpop, emitterPath.slice(0, -1));
+    var event = _.last(emitterPath);
+
+    return _.extend(
+        {},
+        tracerConfig,
+        {
+            extEvent: extEvent,
+            sourceEmitter: emitter,
+            sourceEvent: event,
+        }
+    );
+}
+
+function getIn(obj, path) {
+    for (var i = 0;
+         (typeof obj === 'object' || Array.isArray(obj)) && i < path.length;
+         ++i) {
+        obj = obj[path[i]];
+    }
+    return obj;
+}
+
+
+
+// base for sink-specific tracers
+function Tracer(ringpop, config, opts) {
+    if (!(this instanceof Tracer)) {
+        return new Tracer(opts);
+    }
+
+    this.ringpop = ringpop;
+    _.extend(this, config);
+    this.opts = opts;
+    this.listener = null;
+}
+util.inherits(Tracer, events.EventEmitter);
+
+Tracer.validate = function validate(opts) {
+    return (
+        opts &&
+        opts.sink
+    );
+};
+
+Tracer.prototype.validate = function validate() {
+    return Tracer.validate(this.opts);
+}
+
+Tracer.prototype.connectSource = function connectSource() {
+    if (!this.listener) {
+        this.listener = this.send.bind(this);
+        this.sourceEmitter.on(this.sourceEvent, this.listener);
+    }
+};
+
+Tracer.prototype.disconnectSource = function disconnectSource() {
+    if (this.listener) {
+        this.sourceEmitter.removeListener(this.sourceEvent, this.listener);
+    }
+    this.listener = null;
+};
+
+
+
+module.exports = {
+    resolveTracerConfig: resolveTracerConfig,
```

Comment:
The config is indeed exported, but I disagree on this point -- resolution is an action, not passive config. As previously mentioned it could be loaded during init, but not at compile/load time, so I argue that it's a core func, even if not repeatedly called. Do you buy it? @jwolski